### PR TITLE
New structure for test repositories and added testing suites  

### LINKED ISSUE #66: Edit test repositories to have multiple files and trivial titles
### Existing issue check

- [X] I found no existing [issues](https://github.com/software-gardening/almanack/issues) covering this topic.

### Description

This addition would allow for enhanced testing functionality, as well as proper representation of existing repositories. This looks like having multiple repositories with multiple files in each of them. Each repository should have different amounts/types of changes , ensuring our modules can work on a variety repositories. 

### Example

* Test_repo_1
   - file_1.md
   - file_2.md
   - file_3.md
* Test_repo_2
  - file_1.md
  - file_2.md
* Referencing these comments

  - Consider the use of [`range`](https://docs.python.org/3/library/functions.html#func-range) and [`join`](https://docs.python.org/3/library/stdtypes.html#str.join) to help create a more programmatic string of numbers as you have here.
  
    _Originally posted by @d33bs in [discussion](https://github.com/software-gardening/almanack/pull/65#discussion_r1665788813)_
            
  - Consider adding a comparison for high entropy vs low entropy with an additional assert below (should one be higher than the other)?
  
    _Originally posted by @d33bs in [discussion](https://github.com/software-gardening/almanack/pull/65#discussion_r1665795678)_

  - Consider finding a way to avoid the repetition for the baseline content additions to these files (while retaining the same outcome).

    _Originally posted by @d33bs in https://github.com/software-gardening/almanack/pull/65#discussion_r1665784628_

  - Consider adding another check to make sure that the file sets are different from one another.

    _Originally posted by @d33bs in https://github.com/software-gardening/almanack/pull/65#discussion_r1665797270_
            
            

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by willdavidson05**
# Description
This PR enhances our test repositories to better simulate existing repositories for accurate testing. It also adds new testing suites for the refactored code from the most recent merge. 
#### Current Repository Structure:
- **High_entropy**
  - File_1.md:
    - Baseline content (commit 1)
    - High entropy added (commit 2)
- **Low_entropy**
  - File_1.md:
    - Baseline content (commit 1)
    - Low entropy added (commit 2)
 
#### Improved Repository Structure:
- **Test_repo_1**
  - File_1.md:
    - Baseline content (commit 1)
    - High amount of added change (commit 2)
  - File_2.md:
    - Baseline content (commit 1)
    - Medium amount of added change (commit 2)
  - File_3.md:
    - Baseline content (commit 1)
    - Low amount of added change (commit 2)

- **Test_repo_2**
  - File_1.md:
    - Baseline content (commit 1)
    - Low amount of added change (commit 2)

This updated structure allows for robust testing, specifically with `calculate_normalized_entropy`, which measures changes at a file based level relative to the repository. Implemented through `add_data.py` and `add_entropy.py`, these repositories now feature varying levels of code changes.

In `entropy.py` the functions `calculate_normalized_entropy`,  `aggregate_entropy_calculation` needed testing suites. 

In `repo_helper`, a test suite for `process_repository` is needed, but due to significant changes in the upcoming PR, it's better to implement the test suite there.

I would appreciate any feedback or suggestions!

Closes #66, #64, #77

## What is the nature of your change?

- [X] Content additions or updates (adds or updates content)
- [ ] Bug fix (fixes an issue).
- [ ] Enhancement (adds functionality).
- [ ] Breaking change (these changes would cause existing functionality to not work as expected).

# Checklist

Please ensure that all boxes are checked before indicating that this pull request is ready for review.

- [X] I have read the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines.
- [X] My code follows the style guidelines of this project.
- [X] I have performed a self-review of my own contributions.
- [X] I have commented my content, particularly in hard-to-understand areas.
- [X] I have made corresponding changes to related documentation (outside of book content).
- [X] My changes generate no new warnings.
- [X] New and existing tests pass locally with my changes.
- [X] I have added tests that prove my additions are effective or that my feature works.
- [X] I have deleted all non-relevant text in this pull request template.

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -92,8 +92,23 @@ def repository_paths(tmp_path_factory):
     create_repositories(base_path)
 
     repositories = {
-        "high_entropy": base_path / "high_entropy",
-        "low_entropy": base_path / "low_entropy",
+        "test_repo_1": base_path / "test_repo_1",
+        "test_repo_2": base_path / "test_repo_2",
```
Comment (d33bs) [inline tests/conftest.py:96]: Consider keeping some sort of adjective or descriptive word with these test repo names to help describe how they're different at a glance. Maybe something about multi- vs single file?

File: tests/conftest.py [hunk 2]
```diff
@@ -92,8 +92,23 @@ def repository_paths(tmp_path_factory):
...
     }
 
     yield repositories
+
+
+@pytest.fixture
+def test_file_sets():
```
Comment (d33bs) [inline tests/conftest.py:103]: Consider adding to the name here about "which" file set (repo?). Also, consider avoiding the prefix `test_` to help avoid this being interpreted as a test vs fixture.

File: tests/data/almanack/entropy/add_data.py [hunk 3]
```diff
@@ -9,56 +9,67 @@
...
+        Repository Structure:
+
+            Test_repo_1:
+                - File_1.md (Lines Changed: 61)
+                - File_2.md (Lines Changed: 31)
+                - File_3.md (Lines Changed: 16)
+            Test_repo_2:
+                - File_1.md (Lines Changed: 16)
+
+        The baseline content is committed first, followed by the addition of entropy.
```
Comment (d33bs) [inline tests/data/almanack/entropy/add_data.py:42]: Aligning with the sentence below.
```suggestion
        The baseline content is committed first, followed by additional lines of changes.
```

File: tests/data/almanack/entropy/add_data.py [hunk 4]
```diff
@@ -9,56 +9,67 @@
...
+        elif repo_name == "test_repo_2":
+            md_file = repo_path / "file_1.md"
+            # Add baseline content to the Markdown file
+            with open(md_file, "w") as f:
+                f.write("Baseline content")
+            repo.index.add([str(md_file)])
+            repo.index.commit("Committed baseline content to test_repo_2")
 
-    # Run the add_entropy.py module
+    # Run the add_entropy.py module to insert entropy
```
Comment (d33bs) [inline tests/data/almanack/entropy/add_data.py:69]: Consider keeping things simple here about what's being added.
```suggestion
    # Run the add_entropy.py module to insert additional lines
```

File: tests/data/almanack/entropy/add_data.py [hunk 5]
```diff
@@ -9,56 +9,67 @@
...
     insert_entropy(base_path)
 
     # Commit changes after adding entropy
-    for dir_name in ["high_entropy", "low_entropy"]:
-        with open(base_path / "high_entropy/high_entropy2.md", "w") as f:
-            f.write("List of Number: 1\n,2\n,3\n,4\n,5\n,6\n,7\n8\n,9\n")
-        commit_changes(base_path / dir_name, "Commit with added entropy")
+    for repo_name in ["test_repo_1", "test_repo_2"]:
+        repo_path = base_path / repo_name
+        commit_changes(repo_path, "Commit with added entropy")
```
Comment (d33bs) [inline tests/data/almanack/entropy/add_data.py:75]:
```suggestion
        commit_changes(repo_path, "Commit with added lines")
```

File: tests/data/almanack/entropy/add_entropy.py [hunk 6]
```diff
@@ -18,69 +18,38 @@ def add_entropy_to_file(file_path: pathlib.Path, lines_of_code: str) -> None:
...
-
-
-low_entropy_code = """
-        ## Simple List
-        - Item 1
-        - Item 2
-        - Item 3
-        - Item 4
-        - Item 5
+    * 20
```
Comment (d33bs) [inline tests/data/almanack/entropy/add_entropy.py:25]: Nice!

File: tests/data/almanack/entropy/add_entropy.py [hunk 7]
```diff
@@ -18,69 +18,38 @@ def add_entropy_to_file(file_path: pathlib.Path, lines_of_code: str) -> None:
...
     """
+\n
+"""
+    * 5
+)
 
 
 def insert_entropy(base_path: pathlib.Path) -> None:
+    """
+    Inserts entropy into the specified files in the test repositories.
```
Comment (d33bs) [inline tests/data/almanack/entropy/add_entropy.py:43]: Consider also addressing the function name and other content here related to this change.
```suggestion
    Inserts lines into the specified files in the test repositories.
```

**Review by d33bs — COMMENTED** [review]
Nice job! Overall I thought things were looking good, I dig the new file tests. I left a few comments a suggestions up for your consideration.

Additionally: there might be a chance to perform some ["opportunistic refactoring"](https://martinfowler.com/bliki/OpportunisticRefactoring.html) in moving from `gitpython` to `pygit2` usage within the tests. This could have benefits in terms of maintenance understanding and also test speed. It could also be future work, depending on what you think.

**Comment by willdavidson05** [comment]
@d33bs, thanks for the review! I have implemented most of your feedback. I agree that it makes more sense to stray away from using the word "entropy" in instances where LOC are being implemented. 

I have also made changes to the folder and file names. Please let me know what you think!

I also think that refactoring is needed, since we made the switch to pygit2. I will reference that comment in a new issue, and create a separate PR.

**Review by d33bs — APPROVED** [review]
LGTM! Cheers, nice work!

**Pull request merged**